File: cpp/src/poly_dyn.cpp
#include "utils.cpp"
#include <array>
#include <benchmark/benchmark.h>
#include <chrono>
#include <cstddef>
#include <fstream>
#include <linux/prctl.h>
#include <print>
#include <sys/prctl.h>

enum class SoundEnum { Woof, Meow, Squeek };

struct AnimalInterface {
  virtual SoundEnum sound() const = 0;
  virtual ~AnimalInterface() = default;
};

struct Dog : public AnimalInterface {
  uint64_t id = 1;

  SoundEnum sound() const override {
    if (id == 0)
      return SoundEnum::Meow;
    return SoundEnum::Woof;
  }
};

struct Cat : public AnimalInterface {
  uint64_t id = 1;

  SoundEnum sound() const override {
    if (id == 0)
      return SoundEnum::Woof;
    return SoundEnum::Meow;
  }
};

struct Mouse : public AnimalInterface {
  uint64_t id = 1;

  SoundEnum sound() const override {
    if (id == 0)
      return SoundEnum::Woof;
    return SoundEnum::Squeek;
  };
};

int main() {
  const std::size_t SIZE = 500;
  const std::size_t ITERS = 1000;

  std::ofstream perf_ctl("/tmp/perf.ctl");
  std::ifstream perf_ack("/tmp/perf.ack");

  std::array<SoundEnum, SIZE * ITERS> sound_outputs;
  std::array<SoundEnum, SIZE * ITERS> sound_outputs_warmup;

  // Single stack instances
  Dog dog;
  Cat cat;
  Mouse mouse;

  std::array<AnimalInterface *, SIZE> zoo;

  // Load animal array from file
  std::array<int, SIZE> input_arr =
      utils::read_array_from_file<int, SIZE>("animals.txt");

  for (std::size_t i = 0; i < SIZE; ++i) {
    switch (input_arr[i]) {
    case 1:
      zoo[i] = &dog;
      break;
    case 2:
      zoo[i] = &cat;
      break;
    case 3:
      zoo[i] = &mouse;
      break;
    default:
      std::unreachable();
    }
  }

  benchmark::DoNotOptimize(zoo);

  std::size_t ind = 0;

  // 1. ----------- WARMUP -------------
  for (std::size_t j = 0; j < ITERS; j++) {
    for (const auto &el : zoo) {
      sound_outputs_warmup[ind++] = el->sound();
    }
  }

  ind = 0;

  benchmark::DoNotOptimize(zoo);

  // start perf, then the clock
  utils::send_perf_cmd(perf_ctl, perf_ack, "enable");
  auto start_time = std::chrono::steady_clock::now();

  // 2. ----------- BENCHMARK -------------
  for (std::size_t j = 0; j < ITERS; j++) {
    for (const auto &el : zoo) {
      sound_outputs[ind++] = el->sound();
    }
  }

  // end the clock, then stop perf
  auto end_time = std::chrono::steady_clock::now();
  utils::send_perf_cmd(perf_ctl, perf_ack, "disable");

  std::print("\n---  VERIFYING OUTPUTS ---\n");
  for (auto [ind, sound] : std::views::enumerate(sound_outputs)) {
    std::print("Index {}, sound: {}\n", ind, static_cast<int>(sound));
  }

  for (auto [ind, sound] : std::views::enumerate(sound_outputs_warmup)) {
    std::print("Index {}, sound_warmup: {}\n", ind, static_cast<int>(sound));
  }

  const auto duration = (end_time - start_time).count();
  std::print("Processed in: [{}] ns\n", duration);

  benchmark::DoNotOptimize(sound_outputs);
}

File: rust/src/poly_dyn.rs
use std::fs::OpenOptions;
use std::hint::black_box;
use std::time::Instant;
mod utils;

#[derive(Debug, Clone, Copy)]
#[repr(u8)]
enum SoundEnum {
    Woof,
    Meow,
    Squeek,
}

trait Animal {
    fn sound(&self) -> SoundEnum;
}

struct Dog {
    id: u64,
}

impl Dog {
    fn new() -> Self {
        Self { id: 1 }
    }
}
impl Animal for Dog {
    fn sound(&self) -> SoundEnum {
        if self.id == 0 {
            SoundEnum::Meow
        } else {
            SoundEnum::Woof
        }
    }
}

struct Cat {
    id: u64,
}

impl Cat {
    fn new() -> Self {
        Self { id: 1 }
    }
}
impl Animal for Cat {
    fn sound(&self) -> SoundEnum {
        if self.id == 0 {
            SoundEnum::Woof
        } else {
            SoundEnum::Meow
        }
    }
}

struct Mouse {
    id: u64,
}
impl Mouse {
    fn new() -> Self {
        Self { id: 1 }
    }
}
impl Animal for Mouse {
    fn sound(&self) -> SoundEnum {
        if self.id == 0 {
            SoundEnum::Woof
        } else {
            SoundEnum::Squeek
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut perf_ctl = OpenOptions::new().write(true).open("/tmp/perf.ctl")?;
    let mut perf_ack = OpenOptions::new().read(true).open("/tmp/perf.ack")?;

    const SIZE: usize = 500;
    const ITERS: usize = 1000;

    let mut sound_outputs: [SoundEnum; SIZE * ITERS] = [SoundEnum::Woof; SIZE * ITERS];
    let mut sound_outputs_warmup: [SoundEnum; SIZE * ITERS] = [SoundEnum::Woof; SIZE * ITERS];

    // Single stack instances
    let dog = Dog::new();
    let cat = Cat::new();
    let mouse = Mouse::new();

    let mut zoo: [&dyn Animal; SIZE] = [&dog; SIZE];

    // Load animal array from file
    let input_arr: [i32; SIZE] = utils::read_array_from_file::<i32, SIZE>("animals.txt")?;

    for i in 0..SIZE {
        zoo[i] = match input_arr[i] {
            1 => &dog as &dyn Animal,
            2 => &cat as &dyn Animal,
            3 => &mouse as &dyn Animal,
            _ => unreachable!(),
        };
    }

    zoo = black_box(zoo);

    let mut ind: usize = 0;

    // 1. ----------- WARMUP -------------
    for _ in 0..ITERS {
        for animal in zoo.iter() {
            sound_outputs_warmup[ind] = animal.sound();
            ind += 1;
        }
    }

    ind = 0;

    zoo = black_box(zoo);

    // start perf, then the clock
    utils::send_perf_cmd(&mut perf_ctl, &mut perf_ack, "enable")?;
    let start_time = Instant::now();

    // 2. ----------- BENCHMARK -------------
    for _ in 0..ITERS {
        for animal in zoo.iter() {
            sound_outputs[ind] = animal.sound();
            ind += 1;
        }
    }

    // end the clock, then stop perf
    let end_time = Instant::now();
    utils::send_perf_cmd(&mut perf_ctl, &mut perf_ack, "disable")?;

    println!("\n---  VERIFYING OUTPUTS ---");
    for (i, res) in sound_outputs.iter().enumerate() {
        println!("Index {}: {:?}", i, res);
    }

    for (i, res) in sound_outputs_warmup.iter().enumerate() {
        println!("Index {}, sound_warmup: {:?}", i, res);
    }

    let duration = end_time.duration_since(start_time).as_nanos();
    println!("Processed in: [{}] ns", duration);

    black_box(sound_outputs);

    Ok(())
}
